index.js

```diff
@@ -720,30 +624,85 @@ TChannelConnection.prototype.send = function send(options, frame, callback) {
     //  throw new Error('duplicate frame id in flight');
     // }
 
-    frame.header.id = id;
-    frame.header.seq = 0;
+    var id = self.handler.sendRequestFrame({
+        arg1: arg1,
+        arg2: arg2,
+        arg3: arg3
+    }, function requestSent(err) {
+        if (err) {
+            self.logger.warn('failed to send request', {
+                id: id,
+                error: err
+            });
+            // TODO: retry?
+            // TODO: early self.completeOutOp(id, err, null, null)?
+        }
+    });
+
     self.outOps[id] = new TChannelClientOp(
-        options, frame, self.channel.now(), callback);
+        options, self.channel.now(), callback);
     self.pendingCount++;
-    var buffer = frame.toBuffer();
-    return self.socket.write(buffer);
 };
 /* jshint maxparams:4 */
 
+TChannelConnection.prototype.runInOp = function runInOp(handler, options, sendResponseFrame) {
+    var self = this;
+    var id = options.id;
+    self.inPending++;
+    var op = self.inOps[id] = new TChannelServerOp(self,
+        handler, self.channel.now(), options, opDone);
+
+    function opDone(err, res1, res2) {
+        if (self.inOps[id] !== op) {
+            self.logger.warn('attempt to send frame for mismatched operation', {
+                hostPort: self.channel.hostPort,
+                opId: id
+            });
+            return;
+        }
+        sendResponseFrame(err, res1, res2, function responseSent(err) {
+            if (err) {
+                self.logger.warn('failed to send response', {
+                    id: id,
+                    error: err
+                });
+                // TODO: retry
```

Comment:
a response frame id is scoped to a socket. in an error case the socket is dead. so we cant really retry.